File: backend/services/accountingService.js
const { pool, sql } = require('../config/db');

async function getAccounts(companies = [], query = {}) {
  try {
    await pool.connect();
    // Basic implementation: try to read from ERP_ACCOUNTS or fallback to empty
    const companyFilter = companies.length ? `WHERE Company_Id IN (${companies.join(',')})` : '';
    const q = `SELECT * FROM ERP_ACCOUNTS ${companyFilter} ORDER BY AccountCode`;
    const r = await pool.request().query(q);
    return r.recordset || [];
  } catch (err) {
    console.warn('getAccounts: tabla ERP_ACCOUNTS no disponible o error', err.message || err);
    return [];
  }
}

async function createAccount(payload) {
  try {
    await pool.connect();
    // Minimal insert - assumes columns: AccountCode, Name, Type, Company_Id, ParentAccount
    const { AccountCode, Name, Type, Company_Id, ParentAccount } = payload;
    const r = await pool.request()
      .input('AccountCode', sql.VarChar(50), AccountCode)
      .input('Name', sql.VarChar(255), Name)
      .input('Type', sql.VarChar(50), Type)
      .input('Company_Id', sql.Int, Company_Id)
      .input('ParentAccount', sql.VarChar(50), ParentAccount || null)
      .query(`INSERT INTO ERP_ACCOUNTS (AccountCode, Name, Type, Company_Id, ParentAccount)
              VALUES (@AccountCode, @Name, @Type, @Company_Id, @ParentAccount);
              SELECT SCOPE_IDENTITY() AS Account_Id;`);
    return r.recordset && r.recordset[0] ? r.recordset[0] : null;
  } catch (err) {
    console.error('createAccount error', err.message || err);
    throw err;
  }
}

async function getBalances(companies = [], from, to, accountId) {
  try {
    await pool.connect();
    // Generic ledger query - expects ERP_LEDGER with columns: Date, AccountCode, Debit, Credit, Company_Id
    let where = [];
    if (companies.length) where.push(`Company_Id IN (${companies.join(',')})`);
    if (accountId) where.push(`AccountCode = '${accountId}'`);
    if (from) where.push(`Date >= '${from}'`);
    if (to) where.push(`Date <= '${to}'`);
    const whereSql = where.length ? `WHERE ${where.join(' AND ')}` : '';
    const q = `SELECT AccountCode, SUM(ISNULL(Debit,0)) as Debit, SUM(ISNULL(Credit,0)) as Credit
               FROM ERP_LEDGER ${whereSql}
               GROUP BY AccountCode`;
    const r = await pool.request().query(q);
    return r.recordset || [];
  } catch (err) {
    console.warn('getBalances: tabla ERP_LEDGER no disponible o error', err.message || err);
    return [];
  }
}

async function getIncomeStatement(companies = [], from, to) {
  try {
    await pool.connect();
    // Simplified income statement: sum revenues and expenses by account type
    const companyFilter = companies.length ? `AND Company_Id IN (${companies.join(',')})` : '';
    const q = `
      SELECT AccountType,
        SUM(ISNULL(Debit,0)) as TotalDebit,
        SUM(ISNULL(Credit,0)) as TotalCredit
      FROM (
        SELECT a.Type as AccountType, l.Debit, l.Credit, l.Company_Id
        FROM ERP_LEDGER l
        LEFT JOIN ERP_ACCOUNTS a ON a.AccountCode = l.AccountCode
      ) t
      WHERE 1=1 ${companyFilter}
      GROUP BY AccountType`;
    const r = await pool.request().query(q);
    return r.recordset || [];
  } catch (err) {
    console.warn('getIncomeStatement error', err.message || err);
    return [];
  }
}

async function getSupplierInvoiceRelations(companies = [], supplierId, from, to) {
  try {
    await pool.connect();
    let where = [];
    if (companies.length) where.push(`c.Company_Id IN (${companies.join(',')})`);
    if (supplierId) where.push(`c.Supplier_Id = ${supplierId}`);
    if (from) where.push(`c.Date >= '${from}'`);
    if (to) where.push(`c.Date <= '${to}'`);
    const whereSql = where.length ? `WHERE ${where.join(' AND ')}` : '';
    // Attempt to join purchases/compras with ledger entries - adjust to actual schema if needed
    const q = `
      SELECT c.Compra_Id, c.Folio, c.Date, c.Supplier_Id, c.Total,
             l.AccountCode, l.Debit, l.Credit
      FROM ERP_COMPRAS c
      LEFT JOIN ERP_LEDGER l ON l.Reference_Id = c.Compra_Id
      ${whereSql}
      ORDER BY c.Date DESC`;
    const r = await pool.request().query(q);
    return r.recordset || [];
  } catch (err) {
    console.warn('getSupplierInvoiceRelations error', err.message || err);
    return [];
  }
}

async function getOperationalReports(companies = [], params = {}) {
  try {
    await pool.connect();
    const companyIn = companies.length ? `(${companies.join(',')})` : null;

    async function tableExists(name) {
      const res = await pool.request()
        .input('tableName', sql.VarChar, name)
        .query("SELECT COUNT(*) AS cnt FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = @tableName");
      return res.recordset && res.recordset[0] && res.recordset[0].cnt > 0;
    }

    const results = [];

    // Receivables
    if (await tableExists('ERP_ACCOUNTS_RECEIVABLE')) {
      const q = `SELECT SUM(ISNULL(Balance,0)) as Value FROM ERP_ACCOUNTS_RECEIVABLE ${companies.length ? `WHERE Company_Id IN ${companyIn}` : ''}`;
      const r = await pool.request().query(q);
      results.push({ Metric: 'Receivables', Value: (r.recordset[0] && r.recordset[0].Value) || 0 });
    } else {
      // Fallback: try to compute from ledger/accounts where account type/name suggest receivables
      const q = `
        SELECT SUM(ISNULL(l.Debit,0) - ISNULL(l.Credit,0)) AS Value
        FROM ERP_LEDGER l
        LEFT JOIN ERP_ACCOUNTS a ON a.AccountCode = l.AccountCode
        WHERE (
          LOWER(ISNULL(a.Type,'')) LIKE '%receiv%' OR
          LOWER(ISNULL(a.Type,'')) LIKE '%cxc%' OR
          LOWER(ISNULL(a.Name,'')) LIKE '%por cobrar%' OR
          LOWER(ISNULL(a.Name,'')) LIKE '%cliente%' OR
          LOWER(ISNULL(a.Name,'')) LIKE '%clientes%'
        ) ${companies.length ? `AND l.Company_Id IN ${companyIn}` : ''}`;
      const r = await pool.request().query(q);
      results.push({ Metric: 'Receivables', Value: (r.recordset[0] && r.recordset[0].Value) || 0 });
    }

    // Payables
    if (await tableExists('ERP_ACCOUNTS_PAYABLE')) {
      const q = `SELECT SUM(ISNULL(Balance,0)) as Value FROM ERP_ACCOUNTS_PAYABLE ${companies.length ? `WHERE Company_Id IN ${companyIn}` : ''}`;
      const r = await pool.request().query(q);
      results.push({ Metric: 'Payables', Value: (r.recordset[0] && r.recordset[0].Value) || 0 });
    } else {
      const q = `
        SELECT SUM(ISNULL(l.Credit,0) - ISNULL(l.Debit,0)) AS Value
        FROM ERP_LEDGER l
        LEFT JOIN ERP_ACCOUNTS a ON a.AccountCode = l.AccountCode
        WHERE (
          LOWER(ISNULL(a.Type,'')) LIKE '%pay%' OR
          LOWER(ISNULL(a.Type,'')) LIKE '%cxp%' OR
          LOWER(ISNULL(a.Name,'')) LIKE '%por pagar%' OR
          LOWER(ISNULL(a.Name,'')) LIKE '%proveed%'
        ) ${companies.length ? `AND l.Company_Id IN ${companyIn}` : ''}`;
      const r = await pool.request().query(q);
      results.push({ Metric: 'Payables', Value: (r.recordset[0] && r.recordset[0].Value) || 0 });
    }

    // Cash
    if (await tableExists('ERP_CASH_BALANCES')) {
      const q = `SELECT SUM(ISNULL(Balance,0)) as Value FROM ERP_CASH_BALANCES ${companies.length ? `WHERE Company_Id IN ${companyIn}` : ''}`;
      const r = await pool.request().query(q);
      results.push({ Metric: 'Cash', Value: (r.recordset[0] && r.recordset[0].Value) || 0 });
    } else {
      const q = `
        SELECT SUM(ISNULL(l.Debit,0) - ISNULL(l.Credit,0)) AS Value
        FROM ERP_LEDGER l
        LEFT JOIN ERP_ACCOUNTS a ON a.AccountCode = l.AccountCode
        WHERE (
          LOWER(ISNULL(a.Type,'')) LIKE '%cash%' OR
          LOWER(ISNULL(a.Name,'')) LIKE '%caja%' OR
          LOWER(ISNULL(a.Name,'')) LIKE '%banco%'
        ) ${companies.length ? `AND l.Company_Id IN ${companyIn}` : ''}`;
      const r = await pool.request().query(q);
      results.push({ Metric: 'Cash', Value: (r.recordset[0] && r.recordset[0].Value) || 0 });
    }

    return results;
  } catch (err) {
    console.warn('getOperationalReports error', err.message || err);
    return [];
  }
}

module.exports = {
  getAccounts,
  createAccount,
  getBalances,
  getIncomeStatement,
  getSupplierInvoiceRelations,
  getOperationalReports,
};

File: erp_zoro_python/app/services/accounting_service.py
from __future__ import annotations

from typing import Any, Optional

from sqlalchemy import text

from app.db.session import get_connection, get_transaction


def _build_company_in(companies: list) -> tuple[str, dict]:
    if not companies:
        return "", {}
    placeholders = ", ".join(f":c{i}" for i in range(len(companies)))
    params = {f"c{i}": v for i, v in enumerate(companies)}
    return f"({placeholders})", params


def _table_exists(conn, table_name: str) -> bool:
    row = conn.execute(
        text("SELECT COUNT(*) FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = :t"),
        {"t": table_name},
    ).fetchone()
    return bool(row and row[0] > 0)


def get_accounts(companies: list) -> list[dict[str, Any]]:
    in_clause, params = _build_company_in(companies)
    where = f"WHERE Company_Id IN {in_clause}" if in_clause else ""
    try:
        with get_connection() as conn:
            rows = conn.execute(
                text(f"SELECT * FROM ERP_ACCOUNTS {where} ORDER BY AccountCode"),
                params,
            ).mappings().all()
        return [dict(r) for r in rows]
    except Exception:
        return []


def create_account(payload: dict) -> Optional[dict[str, Any]]:
    with get_transaction() as conn:
        row = conn.execute(
            text("""
                INSERT INTO ERP_ACCOUNTS (AccountCode, Name, Type, Company_Id, ParentAccount)
                OUTPUT INSERTED.*
                VALUES (:code, :name, :type, :company, :parent)
            """),
            {
                "code": payload.get("AccountCode"),
                "name": payload.get("Name"),
                "type": payload.get("Type"),
                "company": payload.get("Company_Id"),
                "parent": payload.get("ParentAccount"),
            },
        ).mappings().first()
    return dict(row) if row else None


def update_account(account_code: str, payload: dict) -> Optional[dict[str, Any]]:
    with get_transaction() as conn:
        row = conn.execute(
            text("""
                UPDATE ERP_ACCOUNTS
                SET Name = :name, Type = :type, ParentAccount = :parent
                OUTPUT INSERTED.*
                WHERE AccountCode = :code
            """),
            {
                "code": account_code,
                "name": payload.get("Name"),
                "type": payload.get("Type"),
                "parent": payload.get("ParentAccount"),
            },
        ).mappings().first()
    return dict(row) if row else None


def get_balances(
    companies: list,
    from_date: Optional[str] = None,
    to_date: Optional[str] = None,
    account_id: Optional[str] = None,
) -> list[dict[str, Any]]:
    in_clause, params = _build_company_in(companies)
    conditions: list[str] = []
    if in_clause:
        conditions.append(f"Company_Id IN {in_clause}")
    if account_id:
        conditions.append("AccountCode = :acct")
        params["acct"] = account_id
    if from_date:
        conditions.append("Date >= :fd")
        params["fd"] = from_date
    if to_date:
        conditions.append("Date <= :td")
        params["td"] = to_date

    where = f"WHERE {' AND '.join(conditions)}" if conditions else ""
    try:
        with get_connection() as conn:
            rows = conn.execute(
                text(f"""
                    SELECT AccountCode,
                           SUM(ISNULL(Debit, 0)) AS Debit,
                           SUM(ISNULL(Credit, 0)) AS Credit
                    FROM ERP_LEDGER {where}
                    GROUP BY AccountCode
                """),
                params,
            ).mappings().all()
        return [dict(r) for r in rows]
    except Exception:
        return []


def get_income_statement(
    companies: list,
    from_date: Optional[str] = None,
    to_date: Optional[str] = None,
) -> list[dict[str, Any]]:
    in_clause, params = _build_company_in(companies)
    conditions = ["1=1"]
    if in_clause:
        conditions.append(f"l.Company_Id IN {in_clause}")
    if from_date:
        conditions.append("l.Date >= :fd")
        params["fd"] = from_date
    if to_date:
        conditions.append("l.Date <= :td")
        params["td"] = to_date

    where = " AND ".join(conditions)
    try:
        with get_connection() as conn:
            rows = conn.execute(
                text(f"""
                    SELECT a.Type AS AccountType,
                           SUM(ISNULL(l.Debit, 0)) AS TotalDebit,
                           SUM(ISNULL(l.Credit, 0)) AS TotalCredit
                    FROM ERP_LEDGER l
                    LEFT JOIN ERP_ACCOUNTS a ON a.AccountCode = l.AccountCode
                    WHERE {where}
                    GROUP BY a.Type
                """),
                params,
            ).mappings().all()
        return [dict(r) for r in rows]
    except Exception:
        return []


def get_supplier_invoice_relations(
    companies: list,
    supplier_id: Optional[str] = None,
    from_date: Optional[str] = None,
    to_date: Optional[str] = None,
) -> list[dict[str, Any]]:
    in_clause, params = _build_company_in(companies)
    conditions: list[str] = []
    if in_clause:
        conditions.append(f"c.Company_Id IN {in_clause}")
    if supplier_id:
        conditions.append("c.Supplier_Id = :sup")
        params["sup"] = supplier_id
    if from_date:
        conditions.append("c.Date >= :fd")
        params["fd"] = from_date
    if to_date:
        conditions.append("c.Date <= :td")
        params["td"] = to_date

    where = f"WHERE {' AND '.join(conditions)}" if conditions else ""
    try:
        with get_connection() as conn:
            rows = conn.execute(
                text(f"""
                    SELECT c.OC_Id AS Compra_Id, c.NumeroOC AS Folio, c.FechaOC AS Date,
                           c.Proveedor_Id AS Supplier_Id, c.Total,
                           l.AccountCode, l.Debit, l.Credit
                    FROM ERP_COMPRA_ORDEN c
                    LEFT JOIN ERP_LEDGER l ON l.Reference_Id = c.OC_Id
                    {where}
                    ORDER BY c.Date DESC
                """),
                params,
            ).mappings().all()
        return [dict(r) for r in rows]
    except Exception:
        return []


def get_operational_reports(companies: list) -> list[dict[str, Any]]:
    in_clause, cparams = _build_company_in(companies)
    results: list[dict[str, Any]] = []

    try:
        with get_connection() as conn:
            for metric, table in [
                ("Receivables", "ERP_ACCOUNTS_RECEIVABLE"),
                ("Payables", "ERP_ACCOUNTS_PAYABLE"),
                ("Cash", "ERP_CASH_BALANCES"),
            ]:
                try:
                    if _table_exists(conn, table):
                        where = f"WHERE Company_Id IN {in_clause}" if in_clause else ""
                        row = conn.execute(
                            text(f"SELECT SUM(ISNULL(Balance, 0)) FROM {table} {where}"),
                            cparams,
                        ).fetchone()
                        value = float(row[0] or 0) if row else 0.0
                    else:
                        # Fallback: derivar desde ERP_LEDGER por tipo de cuenta
                        keyword_map = {
                            "Receivables": ("%receiv%", "%cxc%", "%por cobrar%", "%cliente%"),
                            "Payables": ("%pay%", "%cxp%", "%por pagar%", "%proveed%"),
                            "Cash": ("%caja%", "%banco%", "%cash%"),
                        }
                        kws = keyword_map[metric]
                        kw_conds = " OR ".join(f"LOWER(ISNULL(a.Name,'')) LIKE :kw{i}" for i in range(len(kws)))
                        kw_p: dict = {f"kw{i}": kw for i, kw in enumerate(kws)}
                        if in_clause:
                            kw_p.update(cparams)
                            extra = f"AND l.Company_Id IN {in_clause}"
                        else:
                            extra = ""
                        sign = "Debit" if metric in ("Receivables", "Cash") else "Credit"
                        opp = "Credit" if sign == "Debit" else "Debit"
                        row = conn.execute(
                            text(f"""
                                SELECT SUM(ISNULL(l.{sign},0) - ISNULL(l.{opp},0))
                                FROM ERP_LEDGER l
                                LEFT JOIN ERP_ACCOUNTS a ON a.AccountCode = l.AccountCode
                                WHERE ({kw_conds}) {extra}
                            """),
                            kw_p,
                        ).fetchone()
                        value = float(row[0] or 0) if row else 0.0
                    results.append({"Metric": metric, "Value": value})
                except Exception:
                    results.append({"Metric": metric, "Value": 0.0})
    except Exception:
        pass

    return results
